streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionStore.java

```diff
@@ -170,38 +175,67 @@ public void put(final Windowed<Bytes> key, final byte[] value) {
     }
 
     @Override
-    public KeyValueIterator<Windowed<Bytes>, byte[]> fetch(final Bytes from, final Bytes to) {
+    public KeyValueIterator<Windowed<Bytes>, byte[]> fetch(final Bytes from,
+                                                           final Bytes to) {
         Objects.requireNonNull(from, "from cannot be null");
         Objects.requireNonNull(to, "to cannot be null");
         return findSessions(from, to, 0, Long.MAX_VALUE);
     }
 
-    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
+    private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
+                                    final InternalProcessorContext context) {
         final Bytes binaryKey = cacheFunction.key(entry.key());
         final ProcessorRecordContext current = context.recordContext();
         context.setRecordContext(entry.entry().context());
         try {
             final Windowed<K> key = SessionKeySchema.from(binaryKey.get(), serdes.keyDeserializer(), topic);
             final Bytes rawKey = Bytes.wrap(serdes.rawKey(key.key()));
+            final byte[] newValue = entry.newValue();
             if (flushListener != null) {
-                final AGG newValue = serdes.valueFrom(entry.newValue());
-                final AGG oldValue = newValue == null || sendOldValues ?
-                    serdes.valueFrom(bytesStore.fetchSession(rawKey, key.window().start(), key.window().end())) :
-                    null;
-                if (!(newValue == null && oldValue == null)) {
+                final byte[] oldRawValue;
+                // special case for session stores: we need to send old value even if not requested if new value is null
+                if (sendOldValues || newValue == null) {
```

Comment:
This is related to the bug fix that I have on https://github.com/apache/kafka/pull/6191.

Thinking about this, I'd prefer we merge 6191 first since there are quite some overlapping logic with this one, and rebasing 6191 after merging 6152 could be a much larger effort than 6191 then 6152, wdyt?

Reply:
Works for me.